gcpy/cstools.py

```diff
@@ -692,10 +689,113 @@ def is_cubed_sphere(
+    A cubed-sphere data file has one of the following attributes
+    (1) A dimension named "nf" (GCHP/GEOS diagnostic files)
+    (2) The lat/lon ratio is exactly 6 (GCHP/GEOS checkpoints)
+    """
+    gcpy.util.verify_variable_type(data, (xr.DataArray, xr.Dataset))
+
+    if is_cubed_sphere_diag_grid(data):
+        return True
+    if is_cubed_sphere_rst_grid(data):
+        return True
+    return False
+
+
+def is_cubed_sphere_diag_grid(data):
+    """
+    Determines if a cubed-sphere grid has History file dimensions.
+    (i.e. a dimension named "nf", ak number of grid faces).
+
+    Args:
+    -----
+    data : xarray.DataArray or xarray.Dataset
+        The input data.
+
+    Returns:
+    --------
+    True if the grid has restart dimensions, False otherwise.
+    """
+    if "nf" in data.dims:
+        return True
+    return False
+
+
+def is_cubed_sphere_rst_grid(data):
+    """
+    Determines if a cubed-sphere grid has restart file dimensions.
+    (i.e. lat and lon, with lat = lon*6).
+
+    Args:
+    -----
+    data : xarray.DataArray or xarray.Dataset
+        The input data.
+
+    Returns:
+    --------
+    True if the grid has restart dimensions, False otherwise.
     """
     gcpy.util.verify_variable_type(data, (xr.DataArray, xr.Dataset))
 
-    if "nf" in data.dims:   # nf = number of cubed-sphere faces
+    if "lat" in data.dims:
+        if data.dims["lat"] == data.dims["lon"] * 6:
+            return True
+    if "SPC_" in data.data_vars.keys():
         return True
+    return False
+
 
+def get_cubed_sphere_res(data):
+    """
+    Given a Dataset or DataArray object, returns the number of
+    grid cells on each cubed-sphere grid face.
+
+    Args:
+    -----
+    data : xarray.DataArray or xarray.Dataset
+        The input data.
+
+    Returns:
+    --------
+    cs_res : int
+        The cubed-sphere resolution.  Will return 0 if the data
+        is not placed on a cubed-sphere grid.
+    """
+    gcpy.util.verify_variable_type(data, (xr.DataArray, xr.Dataset))
+
+    if not is_cubed_sphere(data):
+        return 0
+    if is_cubed_sphere_rst_grid(data):
+        return data.dims["lon"]
+    return data.dims["Xdim"]
+
+
+def is_gchp_lev_positive_down(data):
+    """
+    Determines if GCHP data is arranged vertically from the top of the
+    atmosphere downwards or from the surface upwards, according to:
+
+    (1) Checkpoint files:     lev:positive="down
+    (2) Emissions collection: lev:positive="down"
+    (3) Other collections     lev:positive="up"
+
+    Args:
+    -----
+    data : xarray.DataArray or xarray.Dataset
+       The input data
+
+    Returns:
+    --------
+    True if the data is arranged from top-of-atm downwards.
+    False if the data is arranged from the surface upwards.
+    """
+    gcpy.util.verify_variable_type(data, (xr.DataArray, xr.Dataset))
+    
+    if is_cubed_sphere_rst_grid(data):
+        return True
+    if is_cubed_sphere_diag_grid(data) and "Emis" in data.data_vars:
```

Comment:
Does this check if "Emis" is anywhere in any of the data variable names?

Reply:
@lizziel: Oops, you're right.  This would also pick out any variable beginning with "Emis":
```python
>>> data_vars = ["SpeciesConcVV_CO", "EmisCO_Biomass", "BudgetChemistryPBL_CO", "BudgetEmisDryDepFull_CO"]
>>> emis_vars = [var for var in data_vars if "Emis" in var]
>>> print(emis_vars)
['EmisCO_Biomass', 'BudgetEmisDryDepFull_CO']
```
I can try to refine the search (i.e. only look in the Emissions collection or something).

Reply:
Ah, this will work:
```python
>>> data_vars = ["SpeciesConcVV_CO", "EmisCO_Biomass", "BudgetChemistryPBL_CO", 
>>> emis_vars = [var for var in data_vars if "Emis" in var and var.startswith("Emis")]
>>> print(emis_vars)
['EmisCO_Biomass']
```
I can do something like this to only pick out variables starting with Emis.

Reply:
I think this is the correct code:
```python
    if is_cubed_sphere_rst_grid(data):
        return True
    if is_cubed_sphere_diag_grid(data):
        emis_vars = [var for var in data.data_vars if var.startswith("Emis")]
        if len(emis_vars) > 0:
            return True
    return False
```

Reply:
@lizziel: I've confirmed that this works.  Running this code snippet:
```python
    diag_dir = "/n/holyscratch01/jacob_lab/msulprizio/GC/benchmarks/1yr/14.2.0-rc.2/GCHP/FullChem/OutputDir"
 

    ifile = "GEOSChem.Emissions.20190701_0000z.nc4"
    dset = xr.open_dataset(join(diag_dir, ifile))
    print(ifile, is_gchp_lev_positive_down(dset))
    
    ifile = "GEOSChem.SpeciesConc.20190701_0000z.nc4"
    dset = xr.open_dataset(join(diag_dir, ifile))
    print(ifile, is_gchp_lev_positive_down(dset))

    quit()
```
gives the following results:
```console
GEOSChem.Emissions.20190701_0000z.nc4 True
GEOSChem.SpeciesConc.20190701_0000z.nc4 False
```

Reply:
@lizziel: Now fixed in commit 45009436.